x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/KeyToSequences.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.execution.sequence;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Dedicated collection for mapping a key to a list of sequences */
+/** The list represents the sequence for each stage (based on its index) and is fixed in size */
+
+class KeyToSequences {
+
+    private final int listSize;
+    private final Map<SequenceKey, List<SequenceFrame>> keyToSequences;
```

Comment:
Inverting this might simplify the `frame()` function: `List<Map<SequenceKey, SequenceFrame>>`

Reply:
Indeed - my concern is on which layout is more memory efficient. Maps are heavier than lists and for a large number of keys that would get hold by the first stages in a sequences, that would result in a number of maps with a large number of keys.
The access time would be faster as the number of keys gets smaller in each stage but (without benchmarking) I considered that to be less of an issue.
Further experimentation will tell which one would work better.
Having a list for every single key (even if transitory) might prove problematic as well.